### DISCUSSION TITLE
关于solidity中变量数据存储的问题

### DISCUSSION THREAD

**Discussion by Sha777wee** [AmazingAng/WTF-Solidity#479] [Q&A] [answered]
solidity中变量相互赋值的时候究竟是按照什么规则来确定是产生独立副本还是产生引用的？变量类型（值类型或引用类型）、存储位置（storage或memory）、定义变量的位置（状态变量或局部变量）这些好像都会影响？求解答

**Comment by XdpCs** [comment, marked as answer]
| 来源 \ 目标   | `storage`          | `memory`           | `calldata`            |
|---------------|--------------------|--------------------|-----------------------|
| **`storage`** | 引用（共享内存）     | 拷贝（独立副本）     | 不允许直接赋值         |
| **`memory`**  | 拷贝（独立副本）     | 引用（共享内存）     | 不允许直接赋值         |
| **`calldata`**| 拷贝（独立副本）     | 拷贝（独立副本）     | 引用（共享内存）
